core/src/Kokkos_Macros.hpp

```diff
@@ -633,8 +633,9 @@
 #define KOKKOS_ATTRIBUTE_NODISCARD
 #endif
 
-#if defined(KOKKOS_COMPILER_GNU) || defined(KOKKOS_COMPILER_CLANG) || \
-    defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_PGI)
+#if (defined(KOKKOS_COMPILER_GNU) || defined(KOKKOS_COMPILER_CLANG) || \
+    defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_PGI)) && \
+    !defined(KOKKOS_COMPILER_MSVC)
```

Comment:
I assume this is actually with the Microsoft clang version right? It is an interesting question of what to do with general clang detection, since now we have all these vendor clang thingies, which identify as clang as well as the vendor compiler. Would it be better to do KOKKOS_COMPILER_MSVC_CLANG as another identifier for example? Since there are certainly situations where we did something different for MSVC proper, which we don't need to do if it is MSVC_CLANG, and right now it would go the MSVC proper path.

Reply:
This is for Clang version downloaded from llvm.org, running as `clang-cl` by adding `-T llvm` as CMake MSVC toolchain option. I thought about these conflicts, but didn't want to impose new variables without consulting. `&& !defined(KOKKOS_COMPILER_MSVC)` seems as easy fix that won't interfere with cygwin + clang and mingw headers + clang. Also, I believe `MSVC_CLANG` crunches all `MSVC` options and ignores redundant ones.  
Btw I still don't know what the status of `clang-cl` and `CUDA`...